src/lib.rs

```diff
@@ -36,12 +39,21 @@ pub trait GraphObject {
 }
 
 /// A graph  node.
-pub trait Node<N>: GraphObject<Data = N> {}
+pub trait Node<N>: GraphObject<Data = N> {
+    /// Returns the type of this node.
+    fn node_type(&self) -> &types::NodeType;
```

Comment:
This shouldn't be needed - the node already has data

Reply:
This is needed because we are implementing the `Node` _trait_ here, and remember that the whole point of this exercise was to allow test code to swap the concrete `NodeData` for something more flexible/simpler. Therefore, it's user code's responsibility to implement this method correctly 😄 (see for example [my open osrank PR](https://github.com/oscoin/osrank-rs/pull/87/files#diff-a0e64c39037b366c6e86d7aa93feb56cR273), where we have an extra layer of indirection, for example).

Reply:
Oh, then shouldn't it be eg. `node_data(&self) -> Self::Data`? Since `types` is the concrete types.

Reply:
There is a bit of history here -- originally I did code everything under the impression that the only little piece of concrete data we were allowing in the interfaces were the `types::NodeData` and `types::EdgeData`, so that both the registry and osrank could have relied on this during data-agnostic code. For example, this would have allowed us to write a generic algorithm function where you had only a `G: Graph` constraint while being able to pattern match on each node type by the virtue of relying on the (concrete) info that each (abstract) node had such a `node_type` function.

However, for other reasons I had to make our osrank algorithm function partially monomorphic in the `G::Data`, which means that, short story short, I should be able to kill this `node_type` method completely, because we can already grab the data data using the `GraphDataReader` or, at the node-level, simply relying on the `GraphObject.data()` function.

Let me have a play with this.